From 070c94013b217bdba19d21c5a810193b24aab9e9 Mon Sep 17 00:00:00 2001
From: Yonggang Luo <luoyonggang@gmail.com>
Date: Thu, 18 Jun 2026 08:20:23 +0800
Subject: [PATCH 08/10] EGL: track make-current for win32 dispatch tables

- Treat eglMakeCurrent like the WGL make-current entrypoints.
- Wrap eglMakeCurrent so successful calls refresh GL and EGL tables.
- Add the external epoxy_handle_external_eglMakeCurrent() notifier.
- Allocate and free EGL TLS storage with the existing win32 TLS setup.
- Document that external eglMakeCurrent calls must notify epoxy.
---
 README.md             |  4 ++++
 include/epoxy/egl.h   |  1 +
 src/dispatch_common.h | 15 +++++++++++++++
 src/dispatch_egl.c    | 42 ++++++++++++++++++++++++++++++++++++++++++
 src/dispatch_wgl.c    | 16 ++++++++++++++++
 src/gen_dispatch.py   |  1 +
 6 files changed, 79 insertions(+)

diff --git a/README.md b/README.md
index 9d83ed9..c07240f 100644
--- a/README.md
+++ b/README.md
@@ -121,6 +121,10 @@ returns values depend on the context's device and pixel format.  If
 change the device or pixel format), then epoxy needs to be notified of
 the change using the `epoxy_handle_external_wglMakeCurrent()` function.
 
+The same applies to EGL on win32.  If `eglMakeCurrent()` is called from
+outside of epoxy, then epoxy needs to be notified of the change using the
+`epoxy_handle_external_eglMakeCurrent()` function.
+
 The win32 `wglMakeCurrent()` variants are slower than they should be,
 because they should be caching the resolved dispatch tables instead of
 resetting an entire thread-local dispatch table every time.
diff --git a/include/epoxy/egl.h b/include/epoxy/egl.h
index b5f8ee0..6d8cde6 100644
--- a/include/epoxy/egl.h
+++ b/include/epoxy/egl.h
@@ -48,6 +48,7 @@ EPOXY_BEGIN_DECLS
 EPOXY_PUBLIC bool epoxy_has_egl_extension(EGLDisplay dpy, const char *extension);
 EPOXY_PUBLIC int epoxy_egl_version(EGLDisplay dpy);
 EPOXY_PUBLIC bool epoxy_has_egl(void);
+EPOXY_PUBLIC void epoxy_handle_external_eglMakeCurrent(void);
 
 EPOXY_END_DECLS
 
diff --git a/src/dispatch_common.h b/src/dispatch_common.h
index 55388a8..f22b174 100644
--- a/src/dispatch_common.h
+++ b/src/dispatch_common.h
@@ -190,11 +190,26 @@ extern epoxy_resolver_failure_handler_t epoxy_resolver_failure_handler;
 #if USING_DISPATCH_TABLE
 void gl_init_dispatch_table(void);
 void gl_switch_to_dispatch_table(void);
+#if PLATFORM_HAS_EGL
+void egl_init_dispatch_table(void);
+void egl_switch_to_dispatch_table(void);
+#endif
 void wgl_init_dispatch_table(void);
 void wgl_switch_to_dispatch_table(void);
 extern uint32_t gl_tls_index, gl_tls_size;
+#if PLATFORM_HAS_EGL
+extern uint32_t egl_tls_index, egl_tls_size;
+#endif
 extern uint32_t wgl_tls_index, wgl_tls_size;
 
+#if PLATFORM_HAS_EGL
+#define eglMakeCurrent_unwrapped epoxy_eglMakeCurrent_unwrapped
+extern EGLBoolean UNWRAPPED_PROTO(eglMakeCurrent_unwrapped)(EGLDisplay dpy,
+                                                               EGLSurface draw,
+                                                               EGLSurface read,
+                                                               EGLContext ctx);
+#endif
+
 #define wglMakeCurrent_unwrapped epoxy_wglMakeCurrent_unwrapped
 #define wglMakeContextCurrentARB_unwrapped epoxy_wglMakeContextCurrentARB_unwrapped
 #define wglMakeContextCurrentEXT_unwrapped epoxy_wglMakeContextCurrentEXT_unwrapped
diff --git a/src/dispatch_egl.c b/src/dispatch_egl.c
index 4978dea..b4ae6de 100644
--- a/src/dispatch_egl.c
+++ b/src/dispatch_egl.c
@@ -27,6 +27,11 @@
 
 #include "dispatch_common.h"
 
+#if USING_DISPATCH_TABLE
+static bool first_context_current = false;
+static bool already_switched_to_dispatch_table = false;
+#endif
+
 int
 epoxy_conservative_egl_version(void)
 {
@@ -120,3 +125,40 @@ epoxy_has_egl(void)
     return false;
 #endif /* PLATFORM_HAS_EGL */
 }
+
+void
+epoxy_handle_external_eglMakeCurrent(void)
+{
+#if USING_DISPATCH_TABLE
+    if (!first_context_current) {
+        first_context_current = true;
+    } else {
+        if (!already_switched_to_dispatch_table) {
+            already_switched_to_dispatch_table = true;
+            gl_switch_to_dispatch_table();
+            egl_switch_to_dispatch_table();
+        }
+
+        gl_init_dispatch_table();
+        egl_init_dispatch_table();
+    }
+#endif
+}
+
+#if USING_DISPATCH_TABLE
+WRAPPER_VISIBILITY (EGLBoolean)
+WRAPPER(epoxy_eglMakeCurrent)(EGLDisplay dpy,
+                              EGLSurface draw,
+                              EGLSurface read,
+                              EGLContext ctx)
+{
+    EGLBoolean ret = epoxy_eglMakeCurrent_unwrapped(dpy, draw, read, ctx);
+
+    if (ret == EGL_TRUE)
+        epoxy_handle_external_eglMakeCurrent();
+
+    return ret;
+}
+
+PFNEGLMAKECURRENTPROC epoxy_eglMakeCurrent = epoxy_eglMakeCurrent_wrapped;
+#endif
diff --git a/src/dispatch_wgl.c b/src/dispatch_wgl.c
index 7baf130..d203211 100644
--- a/src/dispatch_wgl.c
+++ b/src/dispatch_wgl.c
@@ -109,6 +109,11 @@ DllMain(HINSTANCE dll, DWORD reason, LPVOID reserved)
         wgl_tls_index = TlsAlloc();
         if (wgl_tls_index == TLS_OUT_OF_INDEXES)
             return FALSE;
+#if PLATFORM_HAS_EGL
+        egl_tls_index = TlsAlloc();
+        if (egl_tls_index == TLS_OUT_OF_INDEXES)
+            return FALSE;
+#endif
 
         first_context_current = false;
 
@@ -120,6 +125,10 @@ DllMain(HINSTANCE dll, DWORD reason, LPVOID reserved)
 
         data = LocalAlloc(LPTR, wgl_tls_size);
         TlsSetValue(wgl_tls_index, data);
+#if PLATFORM_HAS_EGL
+        data = LocalAlloc(LPTR, egl_tls_size);
+        TlsSetValue(egl_tls_index, data);
+#endif
 
         break;
 
@@ -130,10 +139,17 @@ DllMain(HINSTANCE dll, DWORD reason, LPVOID reserved)
 
         data = TlsGetValue(wgl_tls_index);
         LocalFree(data);
+#if PLATFORM_HAS_EGL
+        data = TlsGetValue(egl_tls_index);
+        LocalFree(data);
+#endif
 
         if (reason == DLL_PROCESS_DETACH) {
             TlsFree(gl_tls_index);
             TlsFree(wgl_tls_index);
+#if PLATFORM_HAS_EGL
+            TlsFree(egl_tls_index);
+#endif
         }
         break;
     }
diff --git a/src/gen_dispatch.py b/src/gen_dispatch.py
index c9b0aaf..208e2ea 100755
--- a/src/gen_dispatch.py
+++ b/src/gen_dispatch.py
@@ -68,6 +68,7 @@ class GLFunction(object):
             'glBegin',
             'glEnd',
             'wglMakeCurrent',
+            'eglMakeCurrent',
             'wglMakeContextCurrentEXT',
             'wglMakeContextCurrentARB',
             'wglMakeAssociatedContextCurrentAMD',
-- 
2.52.0.windows.1

